clean up the deep link code

ntdiary opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This is a followup PR for #16043.
After we remove the redirection ui in the deep link component, We no longer need to update the `appInstallationCheckStatus` variable. 
We can also remove the call to `Network.post()`.
We need to implement the transition process: oldDot --> newDot --> Desktop app 


### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/17391


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
<img src="https://user-images.githubusercontent.com/8579651/234498066-38dec558-975c-4177-800d-328775deca96.png" width="700px" />

Note: If we can't start OldDot locally, we can open the developer tools and add a breakpoint in the source panel to get the `transition` path, just like the image above.

1. Open the local newDot and login with account A (tab 1).
2. Open the local oldDot and login with the same account (tab 2).
3. Open the local Desktop App and stay logged out.
4. Click on the concierge button at bottom right on oldDot/inbox (tab 2).
5. (Optional) Execute `console.dir(t.replace('https://staging.new.expensify.com', ''))` in console to get the `/transition...` path part, create a new tab and type `localhost:8080` and append the transition path part and then open this url (tab 3).
6. Tap on the 'Open Electron.app' button in the tab 3.
7. Verify that account A is also logged in the desktop app.

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
N/A
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Open staging.new.expensify.com and login with account A (tab 1).
2. Open staging.expensify.com and login with the same account (tab 2).
3. Open the staging Desktop App and stay logged out.
4. Click on the concierge button at bottom right on staging.expensify.com/inbox (tab 2).
5. Tap on the 'Open New Expensify' button in the tab 3.
6. Verify that account A is also logged in the desktop app.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/8579651/234503847-c4790472-ecaf-44ca-936b-9b3d93a3172a.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>
N/A
<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>
N/A
<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>
N/A
<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>
N/A
<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>
N/A
<!-- add screenshots or videos here -->

</details>

ntdiary committed:
clean up the deep link code

tgolen commented on src/libs/actions/Session/index.js:
```diff
@@ -642,6 +621,23 @@ function authenticatePusher(socketID, channelName, callback) {
 }
 
+function getShortLivedAuthToken() {
+    // TODO: If we have any way to read the shortLivedAuthToken and save it, we can also no longer call makeRequestWithSideEffects.
```
I'm not sure what "and save it" implies. We are saving it in Onyx, but that isn't enough? Also, TODOs are not allowed in our code. Instead, you should open a GH issue and point to this line of code and then detail what needs to be done.

tgolen commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -62,7 +74,12 @@ class LogInWithShortLivedAuthTokenPage extends Component {
 
     render() {
-        if (this.props.isTokenValid) {
+        // TODO: Not sure if we should redirect the user to the login page or just show the expiration screen or offline message
```
Same comment as above about TODOs (unless you are meaning to clean up this TODO in this PR)

tgolen reviewed (changes requested):
This looks 💯 times better! Thank you

ntdiary committed:
Add some comments to explain the deep link code

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -96,7 +101,7 @@ class DeeplinkWrapper extends PureComponent {
                 }
                 iframe.parentNode.removeChild(iframe);
-            }, 100);
+            }, 0);
```
Why do actually need this `setTimeout()`? The explanation above is not very satisfying...

>we need to give this iframe some time for it to do what it needs to do

What does it mean?

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -48,6 +48,10 @@ class DeeplinkWrapper extends PureComponent {
             return;
         }
+
+        // We need to clear the old short-lived auth token if it exists.
```
Why do we need to do this?

tgolen replied:
Please update the code comment to explain the "why" and not just "what" the code is doing. Why do we need to clear out the previous one before fetching a new one?

tgolen replied:
I think that's a perfect explanation 👍 Here is how I would write it:

> Since there is no way to know if a previous short-lived authToken is still valid, any previous short-lived authToken must be cleared out and a new one must be fetched so that the popup window will only open when we know the short-lived authToken is valid.

marcaaron commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -74,7 +74,7 @@ class LogInWithShortLivedAuthTokenPage extends Component {
-        // TODO: Not sure if we should redirect the user to the login page or just show the expiration screen or offline message
+        // redirect the user to the login page if there is a sudden disconnection
         if (this.props.network.isOffline) {
             Navigation.navigate();
```
Should we be passing something more explicit here? Calling `navigate()` with no arguments seems strange to me.

marcaaron commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -74,7 +74,7 @@ class LogInWithShortLivedAuthTokenPage extends Component {
 
     render() {
-        // TODO: Not sure if we should redirect the user to the login page or just show the expiration screen or offline message
+        // redirect the user to the login page if there is a sudden disconnection
```
I'm unsure what the exact edge case we're trying to handle is based on this comment. The user hits this page and in between downloading the JS bundle and executing it they somehow become offline? Can you expand on what exactly we are preventing and why we are doing this?

marcaaron replied:
I'm sorry I don't think I fully understood the previous comments. I don't think the desktop app needs to download the js bundle because it's an electron app and should be bundled with the app (IIRC).

>if the network is disconnected at this time

If they are disconnected then how can they access the page at all 🤔 

>Without this condition, the session expired ui will be displayed, if we think this page is also ok, then we can remove this condition.

I think it's fine but the message about the session being expired seems wrong?

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -59,8 +63,9 @@ class DeeplinkWrapper extends PureComponent {
         if (prevProps.session.shortLivedAuthToken || !this.props.session.shortLivedAuthToken) {
```
Are we not handling the errors we were handling before anymore?

marcaaron replied:
Can you remind me what errors we were handling and why they are not necessary anymore? I looked at the linked PR but I'm not seeing anything about errors or error handling (maybe I missed it).

marcaaron reviewed (changes requested):
@ntdiary I am working on a clean up for this in another PR... it would have been nice to have a heads up that you are going to tackle improving this code so that we are not duplicating any effort and more generally agree on a plan for how to do it.

marcaaron commented on src/libs/actions/Session/index.js:
```diff
@@ -642,6 +621,23 @@ function authenticatePusher(socketID, channelName, callback) {
+        'OpenOldDotLink', {shouldRetry: false}, {},
+    ).then((response) => {
+        const {shortLivedAuthToken} = response;
+        Onyx.merge(ONYXKEYS.SESSION, {shortLivedAuthToken});
```
**original context:** https://github.com/Expensify/Expensify/issues/218742

arosiclair replied:
Yeah it looks like this should just be an `API.read` now. We're just listening for the short lived token to change in Onyx instead of blocking on the request.

marcaaron commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -107,6 +124,7 @@ LogInWithShortLivedAuthTokenPage.defaultProps = defaultProps;
 export default compose(
     withLocalize,
     withOnyx({
-        isTokenValid: {key: ONYXKEYS.IS_TOKEN_VALID},
```
Just curious, but what was it doing before and why are we removing it now?

ntdiary committed:
Update some comments to explain the deep link code

ntdiary committed:
Update comment for removing the short-lived auth token

ntdiary committed:
specify the route to redirect

ntdiary committed:
Update comment for removing the short-lived auth token

tgolen commented:
Yeah, I'd like to get this merged. Can you please complete the PR description and checklist? It'd also be nice to have a C+ review/test this.

ntdiary committed:
makeRequestWithSideEffects can be removed in the PR 17364

tgolen commented:
bump @arosiclair for review and checklist https://github.com/Expensify/App/pull/17452#issuecomment-1508218684

tgolen commented:
@arosiclair sorry about the bump, I'm still confused on if we are wanting to merge this or not. I'm going to place a HOLD on it until we decide.

ntdiary committed:
rename TRANSITION_FROM_OLD_DOT to TRANSITION

marcaaron commented:
What's the status of this PR? I'm thinking maybe we can create an issue for this to go over the general the plan then treat it like a normal "Bug" issue so that it doesn't fall off our radar? Does that sound OK to you @tgolen ?

tgolen commented:
I thought the plan was to update this PR with all the necessary fixes. There was some discussion about it over here: https://github.com/Expensify/App/issues/17391#issuecomment-1520268938

That's a good reminder that this can be taken off HOLD (but it still needs updated before it can be reviewed).

marcaaron commented:
Ah hmm, first time seeing that comment. If there's a plan let's put it [here](https://github.com/Expensify/App/issues/17987) to get on the same page?

ntdiary committed:
If the current link is transition from oldDot, we should wait for it to finish

ntdiary committed:
prevent the popup from appearing repeatedly

tgolen commented on src/ROUTES.js:
```diff
@@ -103,7 +103,7 @@ export default {
     REPORT_SETTINGS: 'r/:reportID/settings',
     getReportSettingsRoute: reportID => `r/${reportID}/settings`,
-    TRANSITION_FROM_OLD_DOT: 'transition',
+    TRANSITION: 'transition',
```
I think this was renamed because this route is used for two things now:
1. OldDot --> NewDot
2. NewDot (web) --> NewDot (desktop)

so the name "transition from olddot" didn't make sense anymore. I think `TRANSITION` is a bit ambiguous and could mean many different things which doesn't make it very intuitive.

How about calling it `TRANSITION_BETWEEN_APPS`?

tgolen replied:
@marcaaron Not sure if you saw this or not, but you commented about the same thing and this was my reasoning for the suggestion.

marcaaron replied:
Thanks, maybe you can help me understand this because I'm really struggling to see it.

How does this route allow us to go from NewDot web to NewDot desktop?

And if we are using it for that purpose - maybe we should separate these concerns a bit and using some more specific "components" or "routes" that have a more specific purpose?

tgolen replied:
It happens in this:

```
        if (Str.startsWith(window.location.pathname, Str.normalizeUrl(ROUTES.TRANSITION_BETWEEN_APPS))) {
            App.beginDeepLinkRedirectAfterTransition();
            return;
        }
```

Then `App.beginDeepLinkRedirectAfterTransition();` opens up desktop with a deeplink. My understanding is that there are places where OldDot deeplinks directly to the Desktop app (it seems like this might have started in https://github.com/Expensify/App/issues/16083 or earlier), and the only possible way for that to happen is to use NewDot web as a middle-man.

There is a fairly long discussion about this in the issue: https://github.com/Expensify/App/issues/17391#issuecomment-1515999087

tgolen replied:
> maybe we should separate these concerns a bit and using some more specific "components" or "routes" that have a more specific purpose?

I believe this was also mentioned in the discussion a little bit, but there wasn't a strong push for it. I honestly wouldn't mind separating them out either, but it could require changing some things on OldDot (not too big a deal though probably).

ntdiary committed:
rename TRANSITION to TRANSITION_BETWEEN_APPS

ntdiary committed:
rename variable and explain the reason for the code.

ntdiary committed:
correct the condition of variable hasPopupBeenOpenedBefore

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -21,68 +21,76 @@ const propTypes = {
-                ? CONST.DESKTOP_DEEPLINK_APP_STATE.CHECKING : CONST.DESKTOP_DEEPLINK_APP_STATE.NOT_INSTALLED,
-        };
-        this.focused = true;
+        this.hasPopupBeenOpenedBefore = !this.isMacOSWeb() || CONFIG.ENVIRONMENT === CONST.ENVIRONMENT.DEV;
```
I'm not quite making the connection between being on mac os web or dev and whether the popup has been opened before.

Could we add a brief comment here to explain why these platform and environment checks are necessary?

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -21,68 +21,76 @@ const propTypes = {
 
     componentDidMount() {
-        if (!this.isMacOSWeb() || CONFIG.ENVIRONMENT === CONST.ENVIRONMENT.DEV) {
+        if (this.hasPopupBeenOpenedBefore) {
```
Once this is set to `true` then the `DeepLinkWrapper` popup won't ever happen again even when signing out?

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -21,68 +21,76 @@ const propTypes = {
+        // Since there is no way to know if a previous short-lived authToken is still valid,
+        // any previous short-lived authToken must be cleared out and a new one must be fetched
+        // so that the popup window will only open when we know the short-lived authToken is valid.
+        Session.removeShortLivedAuthToken();
```
I don't really get this. Why don't we just not use it when the component mounts? Why does it need to be "cleared out"? It will get replaced with a new token and we will only ever check it once because of the `this.hasPopupBeenOpenedBefore` flag.

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -21,68 +21,76 @@ const propTypes = {
-        const params = new URLSearchParams();
-        params.set('exitTo', `${window.location.pathname}${window.location.search}${window.location.hash}`);
+        // If the current link is transition from oldDot, we should wait for it to finish,
+        // because we should pass the final path to the desktop app instead of the transition path.
```
Can we update this comment to explain what it means for us to "wait for it to finish"? It isn't obvious to me what that means or why we need to do that.

>we should pass the final path to the desktop app instead of the transition path

I'm pretty confused by this statement too. I don't see anything getting 'passed' here.

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -21,68 +21,76 @@ const propTypes = {
         if (!this.props.session.authToken) {
-            const expensifyDeeplinkUrl = `${CONST.DEEPLINK_BASE_URL}${expensifyUrl.host}/transition?${params.toString()}`;
-            this.openRouteInDesktopApp(expensifyDeeplinkUrl);
+            this.openRouteInDesktopApp();
```
Should this be higher up in the order of operations? 

I'm confused because it looks like we early return here if we don't have an `authToken`. Which makes sense because we "get one" on the next line. But the previous line is trying to "get" a short lived token as well.

marcaaron commented on src/components/DeeplinkWrapper/index.website.js:
```diff
@@ -21,68 +21,76 @@ const propTypes = {
+        }
+
+        // Once the popup has appeared after the page is loaded, it does not need to be displayed again.
+        // e.g. When a user refresh one tab, the popup should not appear again in other tabs.
```
I don't understand this comment. Why are tabs relevant in this context? Why would refreshing a tab make popups appear in other tabs?

marcaaron commented on src/libs/Navigation/Navigation.js:
```diff
@@ -313,6 +318,14 @@ function drawerGoBack(backRoute) {
     navigate(backRoute);
 }
 
+function isTransitionEnd() {
```
needs a JS doc

marcaaron commented on src/libs/Navigation/Navigation.js:
```diff
@@ -313,6 +318,14 @@ function drawerGoBack(backRoute) {
+    return transitionIsEndPromise;
+}
+
+function setIsTransitionEnd() {
```
also needs a JS doc

marcaaron commented on src/libs/actions/Session/index.js:
```diff
@@ -266,29 +261,13 @@ function signInWithShortLivedAuthToken(email, authToken) {
-        }
-        Navigation.navigate();
-    });
+    API.read('SignInWithShortLivedAuthToken', {authToken, oldPartnerUserID, shouldRetry: false}, {optimisticData, successData, failureData});
```
I don't think we need the `shouldRetry` here do we? Read requests are not retryable by default IIRC and we have an open issue to remove this param entirely.

marcaaron reviewed (dismissed):
I appreciate the efforts to improve this. But have to admit I'm pretty unsatisfied still with the general readability of this code. It feels like it's below the standards of what we should be shooting for.

That said, I'm struggling to make any great suggestions as I find a lot of it just confusing in general and unclear. 

Sounds like maybe we will have another opportunity to improve it further in the Apple/Google SSO projects. So maybe we can improve some of the documentation then.

ntdiary committed:
Refactor the deeplink logic:

clean up the DeepLinkWrapper/index.website.js
move openRouteInDesktopApp from DeepLinkWrapper to Browser lib.
rename getShortLivedAuthToken/getShortLivedAuthTokenAfterTransition to beginDeepLinkRedirect/beginDeepLinkRedirectAfterTransition and move them to App.js

tgolen commented on src/pages/LogInWithShortLivedAuthTokenPage.js:
```diff
@@ -35,12 +37,23 @@ const propTypes = {
+    }),
+
+    /** Props to detect online status */
+    network: networkPropTypes,
```
It looks like everywhere else in the app it uses `.isRequired` for this prop. Let's do that here.

tgolen commented on src/libs/Browser/index.web.js:
```diff
@@ -50,8 +51,42 @@ function isMobileSafari() {
 }
 
+function openRouteInDesktopApp(shortLivedAuthToken = '', email = '') {
+    const params = new URLSearchParams();
```
Can you please add a code comment here explaining _why_ the params have to be passed this way?

tgolen replied:
Oh, sorry... I am just thinking of a comment like this:

> The session information needs to be passed to the Desktop app, and the only way to do that is by using query params. There is no other way to transfer the data.

tgolen reviewed (changes requested):
OK, taking another look at this since Marc was pulled onto other things. I think there are still a few things which can be cleaned up here to make it a little easier to understand what's going on. Thanks for your continued work on this!

ntdiary committed:
try to improve the transition logic.

tgolen commented on src/libs/Browser/index.web.js:
```diff
@@ -50,8 +51,42 @@ function isMobileSafari() {
+    const expensifyUrl = new URL(CONFIG.EXPENSIFY.NEW_EXPENSIFY_URL);
+    const expensifyDeeplinkUrl = `${CONST.DEEPLINK_BASE_URL}${expensifyUrl.host}/transition?${params.toString()}`;
+
+    // This check is necessary for Safari, otherwise, if the user
+    // does NOT have the Expensify desktop app installed, it's gonna
+    // show an error in the page saying that the address is invalid
```
OK, interesting. Like, what lead you to try an iframe in the first place?

tgolen replied:
It seems like someone had a similar problem here and there was a slightly different way of solving it. Could we try it out? https://stackoverflow.com/questions/18986200/suppress-safari-cant-open-the-page-because-the-address-is-invalid-custom-app-l

Also, do you know if the link URL matches this scheme? https://github.com/Expensify/App/blob/0b35d7c010157841617867ac3924699168cd9d27/ios/NewExpensify/Info.plist#L28

It looks like there are a few more tips in https://csiandal.medium.com/safari-cannot-open-the-page-because-the-address-is-invalid-2147d8963f89

ntdiary committed:
add a comment for openRouteInDesktopApp

tgolen commented on src/libs/actions/App.js:
```diff
@@ -184,6 +185,23 @@ function reconnectApp() {
+ * This promise is used so that deeplink component know when a transition is end.
+ * This is necessary because we want to begin deeplink redirection after the transition is end.
+ */
+let resolveTransitionIsEndPromise;
```
I like seeing it here in `App.js` a lot better, yeah. I think that makes more sense. My last suggestion would be to once again rename this (and the other related methods and variables) to be more a little more intuitive.

What do you think about `waitForSignonTransitionToFinish()` and `endSignonTransition()`

tgolen commented on src/libs/actions/Policy.js:
```diff
@@ -952,7 +952,7 @@ function createWorkspace(ownerEmail = '', makeMeAdmin = false, policyName = '',
     });
 
-    Navigation.isNavigationReady()
+    return Navigation.isNavigationReady()
```
It's not good to return a promise from an action method (our docs state this and it can lead to bad code chaining if someone else sees the pattern and copies it).

I was thinking of how we could stop doing this, and here is what I think:
1. Create a new method in `App.js` called `CreateWorkspaceAndNavigateToIt()`
2. Generate the optimistic policyID in that new method and pass it to `createWorkspace()` as a parameter
3. Move this navigation logic into that new method and only have the API call here

It feels to me like that would clean up a lot of this awkwardness that we're trying to work around and it keeps all the navigation stuff in `App.js` where it should probably belong (or I could even see using a `PolicyCreationUtils.js` file at some point)

tgolen replied:
> Eh, There are two other places that also need navigation logic.

Well, I was thinking those would just be updated to call `CreateWorkspaceAndNavigateToIt()`. Doesn't that still work like it's supposed to?

I think your alternate approach works too, but I still don't really like seeing the navigation happen inside of `createWorkspace()`. IMO, the only thing that method should be doing is creating a workspace. It should not be doing any navigation, because that is business logic that it is not concerned with.

tgolen reviewed (changes requested):
Looks like there is a conflict in `App.js`. 

Thanks for your dedicated work on this! I think we are getting closer, and I had a few more suggestions today as I was looking at the new solution.

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
move navigation logic from createWorkspace to createWorkspaceAndNavigateToIt

ntdiary committed:
fix lint errors

remove network prop as it is unreliable in Firefox, see #17630

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

tgolen commented on src/libs/actions/App.js:
```diff
@@ -195,6 +196,43 @@ function reconnectApp() {
+}
+
+/**
+ * create a new workspace and navigate to it
```
```suggestion
 * Create a new workspace and navigate to it
```

tgolen reviewed (changes requested):
Thanks for those changes. I think this is ready to merge, but I noticed this one last nit-pick. 

I think we can leave the iframe stuff in there, but I'd love to get rid of it at some point if it's not really necessary. Seems like just a bad hack, not your fault though!

ntdiary committed:
Update src/libs/actions/App.js

Co-authored-by: Tim Golen <tgolen@gmail.com>

arosiclair commented:
Sounds like this is ready for the review checklist?

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

tgolen commented:
Yep, last thing here is to get the C+ review and checklist. Thanks for doing this one @0xmiroslav

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

tgolen commented:
I'm looking into that error on the server side. I see it happening for many people

tgolen commented:
The reason the error happens is because the `authToken` that's used to get the short-live-auth-token has expired. So... something must not be refreshing the `authToken` properly. We're still looking into it

cc @yuwenmemon @robertjchen @NikkiWines @johnmlee101

tgolen commented:
There are multiple layers of our code which the error comes from:

- Auth (our database layer) checks the authToken and throws a 407
- PHP (the web API layer) catches the 407 and returns a 666 to the frontend

I'm still trying to get guidance from @NikkiWines and @johnmlee101 about this and bumping them today (including this bump 😁 )

tgolen commented:
OK, so the current running theory is this:

- The `authToken` being sent to the API command 'OpenOldDotLink' is the authToken that's expired (I think this is kind of a "no duh" situation)

I think we need to figure out why that authToken is expired and where it's coming from

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

tgolen commented:
I cannot reproduce the 666 error either. So, it looks like there are a few things left to do here yet:

1. Fix the lint error that's failing
2. Complete the changes mentioned in https://github.com/Expensify/App/pull/17452/files/d9a3cbc67f61d65823bd6384976e090fa97a0404#r1183808816 which were never fully implemented

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Revert the revert regarding createWorkspaceAndNavigateToIt.

ntdiary committed:
fix lint error

tgolen reviewed (approved):
> Do you prefer to have CreateWorkspaceAndNavigateToIt return a Promise? 😂

No, it shouldn't return a promise.

1. Action methods shouldn't ever return promises (there are SUPER rare exceptions)
2. The promise isn't used anywhere
3. It just adds something to the code which the next engineer will see and think it's OK to return a promise in the method they are creating

tgolen commented on src/libs/actions/App.js:
```diff
@@ -213,6 +214,45 @@ function reconnectApp() {
+ * @param {Boolean} [makeMeAdmin] Optional, leave the calling account as an admin on the policy
+ * @param {String} [policyName] Optional, custom policy name we will use for created workspace
+ * @param {Boolean} [transitionFromOldDot] Optional, if the user is transitioning from old dot
+ * @returns {Promise}
+ */
+function createWorkspaceAndNavigateToIt(ownerEmail = '', makeMeAdmin = false, policyName = '', transitionFromOldDot = false) {
+    const policyID = Policy.generatePolicyID();
+    Policy.createWorkspace(ownerEmail, makeMeAdmin, policyName, policyID);
+    return Navigation.isNavigationReady()
```
```suggestion
 */
function createWorkspaceAndNavigateToIt(ownerEmail = '', makeMeAdmin = false, policyName = '', transitionFromOldDot = false) {
    const policyID = Policy.generatePolicyID();
    Policy.createWorkspace(ownerEmail, makeMeAdmin, policyName, policyID);
    Navigation.isNavigationReady()
```

ntdiary committed:
Update src/libs/actions/App.js

Co-authored-by: Tim Golen <tgolen@gmail.com>

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
casing mistake

ntdiary committed:
Do not call clearAccountMessages on non-home screens.

tgolen commented on src/pages/signin/LoginForm.js:
```diff
@@ -75,6 +78,9 @@ class LoginForm extends React.Component {
             login: '',
         };
 
+        if (!this.props.isFocused) {
```
Please add a code comment here for why this logic is here.

tgolen reviewed (changes requested):
For skipping the reauthentication logic, I think you can specify `shouldRetry: false` in the data.

https://github.com/Expensify/App/blob/cd32eccc8768262a18beafb5ec7c0e1be2249111/src/libs/Middleware/Reauthentication.js#L58-L61

ntdiary committed:
add the comment for isFocused

tgolen commented:
Of all the alternatives, I think adding a `skipReauthentication` param is probably the best idea. Just be very verbose with code comments about why it's being added. This solution is better than:

- if the command of the request is SignInWithShortLivedAuthToken, return early. (because `skipReauthentication` can be used on any API request)
- Is it possible to consider adding a shouldReauth? (this is kinda the same thing but a slight different name)
- should we temporarily revert to makeRequestWithSideEffects (because this is going backwards and we want to avoid this)
- the server can return other error codes instead of 407. (because we want to keep the server errors consistent)

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
add a skipReauthentication param

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

s77rt commented:
I will take it

s77rt commented:
Looks we may already have another solution for https://github.com/Expensify/App/issues/17724 ([comment](https://github.com/Expensify/App/issues/17724#issuecomment-1569469641)).
___

@ntdiary Can't we add the "show loader if route is X" logic on top of the current PR code?

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

s77rt commented:
@ntdiary I guess the plan is to wait for https://github.com/Expensify/App/pull/19819 and revert https://github.com/Expensify/App/pull/19125. Is that correct?

Also can you please confirm that the issue https://github.com/Expensify/App/issues/17724 will no longer be reproducible (create a new account)

s77rt commented:
Yeah I got assigned by accident :sweat_smile:. @0xmiroslav is the primary C+ here, I'm just here for "extra" review. Looks like the PR is almost done anyway.

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
we still need the currentUserEmail variable.

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

pecanoro commented:
@0xmiroslav Let us know when you had a chance to take a look!

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
keep variable name consistent.

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

pecanoro commented:
@0xmiroslav Friendly bump to review the last changes asap!

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

pecanoro commented on src/libs/Browser/index.web.js:
```diff
@@ -60,4 +61,43 @@ function isMobileChrome() {
+function openRouteInDesktopApp(shortLivedAuthToken = '', email = '') {
+    const params = new URLSearchParams();
+    params.set('exitTo', `${window.location.pathname}${window.location.search}${window.location.hash}`);
+    if (email && shortLivedAuthToken) {
```
So this means that either both of them are passed or none of them are?

pecanoro commented:
@marcaaron @arosiclair In general, it looks good to me but I am not great with the transitioning logic so it would be great if you guys can give it a review as well!

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

arosiclair reviewed (dismissed):
LGTM as well. All yours @marcaaron

pecanoro commented:
Merge conflicts!

marcaaron commented:
Sorry, I gotta pass on the review for this one as I am scrambling to make up other work and really don't want this to be blocked on my review. I tried to unassign myself a while back, but looks like it's not possible after you've left a comment. 🙃 Anyways, I trust you guys got this.

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

ntdiary committed:
Merge branch 'main' into clean-up-deep-link-code

jasperhuangg commented:
Tests well! Gonna check this off :)

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5602375137) to production by https://github.com/jasperhuangg in version: 1.3.42-26 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5625378789) to staging by https://github.com/arosiclair in version: 1.3.44-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5651460665) to production by https://github.com/marcaaron in version: 1.3.44-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

Pujan92 commented on src/libs/actions/App.js:
```diff
@@ -214,23 +261,25 @@ function setUpPoliciesAndNavigate(session) {
-            // We must call goBack() to remove the /transition route from history
-            Navigation.goBack();
-            Navigation.navigate(exitTo);
-        });
+        Navigation.isNavigationReady()
+            .then(() => {
+                // We must call goBack() to remove the /transition route from history
+                Navigation.goBack();
+                Navigation.navigate(exitTo);
+            })
+            .then(endSignOnTransition);
```
We also need to call `endSignOnTransition` when exitTo param isn't available to avoid [issue](https://github.com/Expensify/App/issues/52450)
